### PR TITLE
[Feature](mlu-ops): update core,gtest,tools-perf

### PR DESCRIPTION
**Pull request by PetrelYy**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

Are new test cases added? If so, please post the corresponding generator-PR link here.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [ ] MLU370
  - [ ] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example:
#
# ----------- case0 -----------
# case0
# [Op name                ]: abs
# [Shape                  ]: input.shape=[1024,1024,3,4], output.shape=[1024,1024,3,4]
# [Data type]             ]: float32
# [MLU Hardware Time      ]: 15728 (us)
# [MLU Interface Time     ]: 369.008 (us)
# [MLU IO Efficiency      ]: 0.23275
# [MLU Compute Efficiency ]: 0.5
# [Mlu Workspace Size     ]: -1 (Bytes)
# 
# ----------- case1 -----------
# ...
```

Platform：MLU590
```bash
# ----------- case0 -----------
# ----------- case1 -----------
# ...
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/core/gen_case.h [hunk 1]
```diff
@@ -463,12 +491,12 @@ class PbNode {
...
-                    std::ofstream &case_file, bool shouldDump,
-                    enum DATASTATE data_state);
+                    std::ofstream &case_file, enum DATASTATE data_state);
   void dumpOutputFile();
-  void dumpToFile(bool isFirst, bool valueDump = false);
+  void dumpToFile(bool valueDump = false);
   void printOnScreen();
-  void serialize(bool isFirst = false);
+  void serialize();
+  void debugTensorAddress();  // may be removed when [MLUOPCORE-13133] completed
```
Comment (DanieeelLiu) [inline bangc-ops/core/gen_case.h:499]: [xxx] 这个内容是否去掉？

File: bangc-ops/core/runtime/device.h [hunk 2]
```diff
@@ -64,7 +64,7 @@ inline int32_t getJobLimitCapability(mluOpHandle_t handle) {
 
 /******************************************************************************
  * MLUOP FUNC: getCoreNumOfJobLimitCapability
- * get mlu core number of every single CNRT_FUNC_TYPE with maximum job capacity.
+ * get ipu core number of every single CNRT_FUNC_TYPE with maximum job capacity.
```
Comment (DanieeelLiu) [inline bangc-ops/core/runtime/device.h]: ipu 好像不能用

File: bangc-ops/core/gen_case.cpp [hunk 3]
```diff
@@ -115,10 +115,13 @@ int genCaseModeGet(bool first) {
     if (it != mode_stacks_.end()) {
       auto &mode_stack = it->second;
       // the top of mode_stack store the gen_case mode for current thread
+
+      // TODO(zhangchaojie): redundant first truth check
```
[the remark this answers is not in the document]
Comment (PetrelYy) [reply bangc-ops/core/gen_case.cpp]: 已修改

File: bangc-ops/core/gen_case.h (hunk 1 above)
Comment (PetrelYy) [reply bangc-ops/core/gen_case.h]: 已修改

File: bangc-ops/core/runtime/device.h (hunk 2 above)
Comment (PetrelYy) [reply bangc-ops/core/runtime/device.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -31,10 +31,17 @@
 #define MLUOP_MINOR 9
 #define MLUOP_PATCHLEVEL 0
 
+/*********************************************************************************
+ * deprecate MLUOP_VERSION is not recommended to use, it is recommended to directly
+ * use MLUOP_MAJOR, MLUOP_MINOR and MLUOP_PATCHLEVEL to get mluOp version.
```
Comment (duzekunKTH) [inline bangc-ops/mlu_op.h:36]: 这一段找suhaoming看一下，看起来语法有错误。

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
  */
 typedef struct mluOpTensorStruct *mluOpTensorDescriptor_t;
 
+/*! The descriptor of Sequence Data that holds the dimensions,
+ * layout, data type, sequence length, padding fill, position, and scale.
+ * The total size of the tensor descriptor only support less than 2 Giga-elements.
+ * You need to call the ::mluOpCreateSeqDataDescriptor function to create a descriptor, and
+ * call the ::mluOpSetSeqDataDescriptor to set the sequence data information to the descriptor.
+ * If the sequence data is in fixed-point data type, call ::mluOpSetSeqDataDescriptorPositionAndScale
```
Comment (duzekunKTH) [inline bangc-ops/mlu_op.h:1011]: 这一整段的字节宽度超过 80 了。从 cnnl 粘出的注释需要根据 mluops的宽度规范更新格式

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+                          int *dimNb,
+                          int dimSize[],
+                          int *seqLengthArraySize,
+                          int seqLengthArray[],
+                          void *paddingFill);
+
+// Group:Common Interface
+// Subgroup:SeqData
+/*!
+ * @brief Retrieves the sequence data descriptor \p seq_data_desc that holds the dimensions,
```
Comment (duzekunKTH) [inline bangc-ops/mlu_op.h:1031]: 此段也超出 80 列了，后面应该也有同样的问题，统一处理一下。

File: bangc-ops/test/mlu_op_gtest/googletest/include/gtest/gtest-death-test.h [hunk 7]
```diff
@@ -257,11 +257,9 @@ class GTEST_API_ KilledBySignal {
 //
 #ifdef NDEBUG
 
-#define EXPECT_DEBUG_DEATH(statement, regex) \
-  GTEST_EXECUTE_STATEMENT_(statement, regex)
+#define EXPECT_DEBUG_DEATH(statement, regex) GTEST_EXECUTE_STATEMENT_(statement, regex)
 
-#define ASSERT_DEBUG_DEATH(statement, regex) \
-  GTEST_EXECUTE_STATEMENT_(statement, regex)
+#define ASSERT_DEBUG_DEATH(statement, regex) GTEST_EXECUTE_STATEMENT_(statement, regex)
```
Comment (duzekunKTH) [inline bangc-ops/test/mlu_op_gtest/googletest/include/gtest/gtest-death-test.h:262]: 88列了，这个能通过格式检查么？或者说从换行改过来是格式刷出来的么？

File: bangc-ops/test/mlu_op_gtest/googletest/include/gtest/gtest-death-test.h [hunk 8]
```diff
@@ -316,23 +314,19 @@ class GTEST_API_ KilledBySignal {
     ::testing::internal::RE::PartialMatch(".*", (regex));                      \
     GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_(statement);                 \
     terminator;                                                                \
-  } else {                                                                     \
-    ::testing::Message();                                                      \
-  }
+  } else                                                                       \
+  ::testing::Message()
```
Comment (duzekunKTH) [inline bangc-ops/test/mlu_op_gtest/googletest/include/gtest/gtest-death-test.h:318]: 这里的修改好奇怪，怎么特地把 else 的代码框+分号删除了？

File: bangc-ops/test/mlu_op_gtest/googletest/include/gtest/gtest-death-test.h [hunk 9]
```diff
@@ -316,23 +314,19 @@ class GTEST_API_ KilledBySignal {
...
-#define EXPECT_DEATH_IF_SUPPORTED(statement, regex) \
-  EXPECT_DEATH(statement, regex)
-#define ASSERT_DEATH_IF_SUPPORTED(statement, regex) \
-  ASSERT_DEATH(statement, regex)
+#define EXPECT_DEATH_IF_SUPPORTED(statement, regex) EXPECT_DEATH(statement, regex)
+#define ASSERT_DEATH_IF_SUPPORTED(statement, regex) ASSERT_DEATH(statement, regex)
 #else
-#define EXPECT_DEATH_IF_SUPPORTED(statement, regex) \
-  GTEST_UNSUPPORTED_DEATH_TEST(statement, regex, )
+#define EXPECT_DEATH_IF_SUPPORTED(statement, regex) GTEST_UNSUPPORTED_DEATH_TEST(statement, regex, )
```
Comment (duzekunKTH) [inline bangc-ops/test/mlu_op_gtest/googletest/include/gtest/gtest-death-test.h:329]: 此处的列宽也超了

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (AquamarineSu) [reply bangc-ops/mlu_op.h:36]: 这部分是要在文档里显示的吗？这么写估计提取不出来呀

确认下是这个意思么？ MLUOP_VERSION is deprecated and not recommended. To get the version of MLUOP, use 
 MLUOP_MAJOR, MLUOP_MINOR and MLUOP_PATCHLEVEL.

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
  */
 typedef struct mluOpTensorStruct *mluOpTensorDescriptor_t;
 
+/*! The descriptor of Sequence Data that holds the dimensions,
+ * layout, data type, sequence length, padding fill, position, and scale.
+ * The total size of the tensor descriptor only support less than 2 Giga-elements.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1008]:
```suggestion
 * The total size of the tensor descriptor supports up to 2 Giga elements.
```

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ * You need to call the ::mluOpCreateSeqDataDescriptor function to create a descriptor, and
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1009]:
```suggestion
 * Call ::mluOpCreateSeqDataDescriptor to create a descriptor, and
```

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ * call the ::mluOpSetSeqDataDescriptor to set the sequence data information to the descriptor.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1010]:
```suggestion
 * call ::mluOpSetSeqDataDescriptor to set the sequence data information to the descriptor.
```

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ * If the sequence data is in fixed-point data type, call ::mluOpSetSeqDataDescriptorPositionAndScale
+ * to set the position and scale of the sequence data.
+ * At last, you need to destroy the descriptor at the end with the ::mluOpDestroySeqDataDescriptor
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1013]:
```suggestion
 * To destroy the descriptor, call ::mluOpDestroySeqDataDescriptor.
```

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ * function.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1014]:
```suggestion
 * 
```

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+                          int dimSize[],
+                          int *seqLengthArraySize,
+                          int seqLengthArray[],
+                          void *paddingFill);
+
+// Group:Common Interface
+// Subgroup:SeqData
+/*!
+ * @brief Retrieves the sequence data descriptor \p seq_data_desc that holds the dimensions,
+ * data type, layout, padding fill and the sequence lengths of the input sequence data.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1032]:
```suggestion
 * data type, layout, padding fill, and the sequence lengths of the input sequence data.
```

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ *
+ * @param[in] seq_data_desc
+ *   Input. The descriptor of the sequence data. For detailed information,
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1035]: 不需要写Input， doxygen语法用@param[in] 表示入参，全文排查一下哦
```suggestion
 *   The descriptor of the sequence data. For detailed information,
```

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ *   see ::mluOpSeqDataDescriptor_t.
+ * @param[out] layout
+ *   Output. The layout of the sequence data. See ::mluOpSeqDataLayout_t.
+ * @param[out] dtype
+ *   Output. The data type of the sequence data. See ::mluOpDataType_t.
+ * @param[out] dimNb
+ *   Output. The number of dimensions of the sequence data.
+ * @param[out] dimSize
+ *   Output. An array that contains the size of the sequence data for each dimension.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1044]:
```suggestion
 *   An array containing the size of the sequence data for each dimension.
```

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ * @param[out] seqLengthArraySize
+ *   Output. Number of elements in sequence length array, \p seqLengthArray[]. It is equal to
+ *   batch * beam (N and B described in the ::mluOpSeqDataLayout_t).
+ * @param[out] seqLengthArray
+ *   Output. An integer array recording the length of all sequences. Note that the array is
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1049]:
```suggestion
 *   An array of  integers recording the length of all sequences. Note that the array is
```

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ *   ordered in a batch-beam order, in despite of sequence data layout. Return NULL when sequence
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1050]:
```suggestion
 *   ordered in a batch-beam order, regardless of the sequence data layout. Return \p NULL when the sequence
```

File: bangc-ops/mlu_op.h [hunk 20]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ *   length array is not set.
+ * @param[out] paddingFill
+ *   Output. A host pointer to the data type \p dtype to fill up the padding vectors within
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1053]:
```suggestion
 *   A host pointer to the data type \p dtype to fill the padding vectors within
```

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ *   the valid length of each sequence. Use NULL when extra padding is not requested.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1054]:
```suggestion
 *   the valid length of each sequence. Use NULL when extra padding is not required.
```

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -970,6 +1003,82 @@ mluOpGetAtomicsMode(mluOpHandle_t handle, mluOpAtomicsMode_t *atomics_mode);
...
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par API Dependency
+ * - Before calling this function, ::mluOpCreateSeqDataDescriptor, ::mluOpSetSeqDataDescriptor_v2
+ *   should be called.
+ *
+ * @note
+ * - dimSize[0] represents the highest dimension, and dimSize[dimNb - 1] represents
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:1063]:
```suggestion
 * - ``dimSize[0]`` represents the highest dimension, and ``dimSize[dimNb - 1]`` represents
```

File: bangc-ops/mlu_op.h [hunk 23]
```diff
@@ -2006,6 +2115,61 @@ mluOpStatus_t MLUOP_WIN_API
...
+// Subgroup:Tensor
+/*!
+ * @brief Retrieves the pointer mode of the input tensor descriptor \p desc set by
+ * ::mluOpSetTensorDescriptorPointerMode.
+ *
+ * @param[in] desc
+ *   Input. The descriptor of the tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[out] pointer_mode
+ *   Input. Pointer to the host memory that holds information about the pointer mode of the input tensor.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:2155]:
```suggestion
 *   Pointer to the host memory holding information about the pointer mode of the input tensor.
```

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -2006,6 +2115,61 @@ mluOpStatus_t MLUOP_WIN_API
 mluOpGetTensorDescriptor(
     const mluOpTensorDescriptor_t desc, mluOpTensorLayout_t *layout, mluOpDataType_t *dtype, int *dimNb, int dimSize[]);
 
+// Group:Common Interface
+// Subgroup:Tensor
+/*!
+ * @brief Sets the pointer mode \p pointer_mode factor to the input tensor descriptor \p desc.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:2121]: 再斟酌一下介词，感觉这里应该用for
```suggestion
 * @brief Sets the pointer mode \p pointer_mode factor for the input tensor descriptor \p desc.
```

File: bangc-ops/mlu_op.h [hunk 25]
```diff
@@ -2006,6 +2115,61 @@ mluOpStatus_t MLUOP_WIN_API
...
+ * @param[in,out] desc
+ *   Input. The descriptor of the tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[in] pointer_mode
+ *   Input. The pointer mode of the input tensor. For detailed information, seee ::mluOpPointerMode_t.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @note
+ * - Currently. the \p pointer_mode setting to MLUOP_POINTER_MODE_HOST is only supported when the number
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:2132]:
```suggestion
 * - Currently, \p pointer_mode setting to MLUOP_POINTER_MODE_HOST is only supported when the number
```

File: bangc-ops/test/mlu_op_gtest/googletest/include/gtest/gtest-death-test.h (hunk 8 above)
Comment (PetrelYy) [reply bangc-ops/test/mlu_op_gtest/googletest/include/gtest/gtest-death-test.h]: 已修复

File: bangc-ops/mlu_op.h [hunk 26]
```diff
@@ -129,8 +136,31 @@ typedef enum {
...
+  MLUOP_LAYOUT_NCL = 10,
+  /*!< The data layout is in the following order: batch size, channel, length.*/
 } mluOpTensorLayout_t;
 
+/******************************************************************************
+ * Cambricon MLUOP sequence data Layout
+ ******************************************************************************/
+//! @brief
+/*! Enumeration variables describing the sequence data (SeqData) layouts. */
+/*! N is batch, B is beam, T is sequence length, C is embedding size. */
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:148]:
```suggestion
 * N represents batch, B represents beam, T represents sequence length, and C represents embedding size. */
```

File: bangc-ops/mlu_op.h [hunk 27]
```diff
@@ -129,8 +136,31 @@ typedef enum {
...
   /*!< The data layout is in the following order: batch size, width, channel. */
+  MLUOP_LAYOUT_NCL = 10,
+  /*!< The data layout is in the following order: batch size, channel, length.*/
 } mluOpTensorLayout_t;
 
+/******************************************************************************
+ * Cambricon MLUOP sequence data Layout
+ ******************************************************************************/
+//! @brief
+/*! Enumeration variables describing the sequence data (SeqData) layouts. */
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:147]:
```suggestion
 * @brief Enumeration variables describing the sequence data layouts. 
```

**Review by AquamarineSu — APPROVED** [review]

**Review by DanieeelLiu — APPROVED** [review]

File: bangc-ops/core/cnlog.hpp [hunk 28]
```diff
@@ -136,7 +136,7 @@ class LogMessage {
   std::string module_name_;       // used to print module name stamp
   bool is_print_head_;            // whether print log head or not
   bool is_print_tail_;            // whether print log tail or not
-  bool is_clear_endl_;            // whether clear endl int the string context
+  bool is_clear_endl_;            // whether clear endl int the string contex
```
Comment (chqy99) [inline bangc-ops/core/cnlog.hpp:139]: 这个单词为什么改

File: bangc-ops/core/context.cpp [hunk 29]
```diff
@@ -227,10 +269,17 @@ mluOpStatus_t MLUOP_WIN_API mluOpCreate(mluOpHandle_t *handle) {
       device_name);  // warning: possible return unknown.
   ctx->sram_size = sram_size - REM_FOR_STACK;
 
+  strncpy(ctx->device_name, device_name, sizeof(device_name));
+  memory_band_width = double(memory_bus_width) * double(memory_clock_rate) /
+                      1000.0 / 1000.0 / 8.0 * 2.0;  // NOLINT
```
Comment (chqy99) [inline bangc-ops/core/context.cpp:274]: 这些数字有什么涵义吗，可以化简吗，/8.0 * 2.0是/4.0吧

File: bangc-ops/core/subscriber.cpp [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+#include "logging.h"
+#include "subscriber.hpp"
+
+namespace mluop {
+namespace pubsub {
+
+size_t Publisher::subscribe(EventType event,
+                            std::function<void(const void *, void *)> handler,
+                            void *usr) {
+#if 0
```
Comment (chqy99) [inline bangc-ops/core/subscriber.cpp:36]: 这块相当于注释？

File: bangc-ops/kernels/tensor_stride_process/tensor_stride_process_host.h [hunk 31]
```diff
@@ -22,36 +22,43 @@
...
-  int total_num = 1;
+  int64_t tensor_dims[MLUOP_DIM_MAX];
+  int64_t tensor_strides[MLUOP_DIM_MAX];
+  uint64_t total_num = 1;
+  uint64_t total_stride = 1;
   bool is_contiguous = 1;
 };
 
-bool ifNeedTensorStrideProcess(const mluOpTensorDescriptor_t tensor_desc);
+bool strideCaseWithNotConsistentDense(int tensor_num, ...);
```
Comment (chqy99) [inline bangc-ops/kernels/tensor_stride_process/tensor_stride_process_host.h:44]: 为什么要调换函数声明顺序，之前字母序i在s前面

**Review by mahxn0 — APPROVED** [review]

File: bangc-ops/core/subscriber.cpp (hunk 30 above)
Comment (PetrelYy) [reply bangc-ops/core/subscriber.cpp:36]: 大部分是调试代码，同步后比较多，不一一修改了

File: bangc-ops/core/context.cpp (hunk 29 above)
Comment (PetrelYy) [reply bangc-ops/core/context.cpp:274]: 带宽计算，不明白8 2 含义如何，可以找 owner 具体确认下。 /8 * 2 并不等价 * 4

File: bangc-ops/core/cnlog.hpp (hunk 28 above)
Comment (PetrelYy) [reply bangc-ops/core/cnlog.hpp]: 同步的注释错误，遗漏了此处，已修改

**Review by AquamarineSu — APPROVED** [review]

**Pull request closed without merging**